### ISSUE TITLE
systemd.services.<foo>.path and buildInput hooks?

### ISSUE THREAD

**Issue by copumpkin** [NixOS/nixpkgs#22498]
Today, it can be a bit jarring for people to add something like `pythonPackages.foo` to `systemd.services.<foo>.path`, expecting it to behave like `buildInputs` in a build. Perhaps it'd be nice to have a `systemd.services.<foo>.inputs` option that works like `buildInputs` and triggers hooks that would make packages with hooks that modify the environment work properly?

**Comment by abbradar** [comment]
That's a bit strange. Why would anyone want to have a build environment (`setup.sh` and stuff) in a systemd service?

**Comment by copumpkin** [comment]
I'm not saying they would, but we often merge build and "actually work"
functionality in there. Before we added python.withPackages a little while
ago, how would you add a python package to the scope of a systemd service?
The hook adds to the PYTHONPATH.

**Comment by abbradar** [comment]
Hm, seems that we do the wrong thing :D I'm not sure how to do better though; in this particular case I'd set `systemd.services.foo.environment.PYTHONPATH` but that's non-ideal too...

**Comment by copumpkin** [comment]
Yeah, this feels like part of the broader build vs. run question. It also arises when we talk about the uses for `nix-shell` as a general-purpose dev environment vs. as a mechanism for debugging nix builds. It feels like we should start talking about how to separate things like `PYTHONPATH` which are useful for execution and other things like autoreconf buildhooks that are only used for building.

cc @edolstra since there was a lot of talk about the new Nix command-line interface and the various uses for `nix-shell`.

**Comment by copumpkin** [comment]
Another example here is `JAVA_HOME`. I had someone get quite confused that `java` was in the `PATH`, but `JAVA_HOME` was unset. If you just use `nix-shell -p jre` you get a proper `JAVA_HOME` in your environment.

**Comment by bennofs** [comment]
I don't think this is a good idea. It conflates the build- and run-time environment. For example, do you really want to pull in `stdenv` and `PKG_CONFIG_DIRS` and your runtime environment for services? That feels wrong to me.

I believe that there are better, different solutions for these problems. For python packages for example, you should not need to explictly set `PYTHONPATH`. Instead, use either `pythonWithPackages` or package your app as a full derivation and then use it in your service. The python builder will make sure that all binaries of your app are wrapped such that `PYTHONPATH` gets set before running them. A similar solution should be possible for `JAVA_HOME`.

**Comment by copumpkin** [comment]
@bennofs I'm not too opinionated on the specific mechanism, but requiring `withPackages`-like solutions further forks the behavior between a build (well tested, every time we run builds) and what people actually use, so I'm hesitant. In other words, why should `buildInputs` be special? Why not use `python.withPackages` there too?

Anyway, as I said I don't care about the specifics, but I am wary of using completely separate mechanisms for the two different use cases.

**Comment by bennofs** [comment]
@copumpkin This is a valid concern. It definitely applies to development environments: it would be nice if we could use the same mechanisms as we use for building the packages. However, there's multiple issues with this:

* a systemd service hould not run a in a "development environment", in my opinion. Services are a runtime construction. This is a lot more clear in the case of compiled languages: you don't need headers or pkgconfig files to run a c-based service, so you shouldn't run the build hooks for this stuff.
* sometimes, the mechanisms for finding dependencies etc used at build time are not really convenient to use if you don't use a script, so they would not be fit for interactive use or even just for use outside of the builder for this language (case in point: haskell at least used to pass explictly flags for every dependency in it's builder)
* using `withPackage`-like wrappers in the build has a cost, it results in more derivations or at least requires time to build the wrappers. These are not such a problem for usage in runtime environments.

But I don't think that a runtime-environment should be the same as a development/build-environment. For example, the issue you described does not only affect systemd units. There's lot's of other places where this would fail as well:

* you cannot place such a package in systemPackages and have it "just work"
* you cannot install such a package with nix-env easily.

I believe just running systemd services inside a build environment is not the solution here. It will perhaps fix this particular issue, but it leaves those other issues unsolved and also is kind of ugly because build environments are often not meant to exists for extended amounts of time (for example, they might create lots of temporary files etc)

**Comment by copumpkin** [comment]
> a systemd service hould not run a in a "development environment", in my opinion.

Oh, I fully agree. I just want things to work reliably 😄 

> you don't need headers or pkgconfig files to run a c-based service, so you shouldn't run the build hooks for this stuff.

No but sometimes I need `(DY)LD_LIBRARY_PATH`. I get that in almost all cases that sort of thing should be handled by a wrapper, but I want to make sure that it's hard to screw up too.

> But I don't think that a runtime-environment should be the same as a development/build-environment.

100% agree 😄 

My point in this ticket is just that there are some things you might want in both contexts, both at runtime and at development/build time. Definitely not proposing bringing in all sorts of annoying build-time dependencies in at runtime; I just don't think we maintain enough information to do this sensibly right now.

**Comment by mmahut** [comment]
Are there any updates to this issue, please?
